### PR TITLE
[PIN-3389] Catalog Management update and upgrade/submission refactor

### PR DESCRIPTION
**Pull request by TonioGela**
- Adapted to new catalog management
- Saving state
- Saving state
- Fixed tests

### CODE REVIEW

File: project/Versions.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ object PagopaVersion {
   lazy val agreementManagementVersion         = "1.0.x-SNAPSHOT"
   lazy val attributeRegistryManagementVersion = "1.0.x-SNAPSHOT"
   lazy val authorizationManagementVersion     = "1.0.x-SNAPSHOT"
-  lazy val catalogManagementVersion           = "1.0.x-SNAPSHOT"
+  lazy val catalogManagementVersion           = "PIN-3377-SNAPSHOT"
```
Comment (galales) [inline project/Versions.scala]: 🔙

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 2]
```diff
@@ -222,7 +249,8 @@ final case class AgreementApiServiceImpl(
       eService       <- catalogManagementService.getEServiceById(agreement.eserviceId)
       _              <- verifyCloningConflictingAgreements(requesterOrgId, agreement.eserviceId)
       consumer       <- tenantManagementService.getTenant(requesterOrgId)
-      _              <- validateCertifiedAttributes(eService, consumer).whenA(eService.producerId != consumer.id)
+      descriptor     <- eService.descriptors.find(_.id == agreement.descriptorId).toFuture(new Exception)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: 🎯

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 3]
```diff
@@ -319,9 +347,10 @@ final case class AgreementApiServiceImpl(
       _              <- assertRequesterIsProducer(requesterOrgId, agreement)
       _              <- agreement.assertRejectableState.toFuture
       eService       <- catalogManagementService.getEServiceById(agreement.eserviceId)
+      descriptor     <- eService.descriptors.find(_.id == agreement.descriptorId).toFuture(new Exception)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: 🎯

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 4]
```diff
@@ -545,33 +578,65 @@ final case class AgreementApiServiceImpl(
...
+                  verifiedAttributes = a.verifiedAttributes,
+                  stamps = a.stamps.copy(archiving = Stamp(uid, offsetDateTimeSupplier.get()).some)
+                )
+              )
+            )
+            .map(_ => ())
+        else Future.unit
+
+      // * Se lo stato non è attivo a causa dei flags ritorna un errore applicativo
+      // * Per questo viene fatto il controllo a valle
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: 🇮🇹

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 5]
```diff
@@ -545,33 +578,65 @@ final case class AgreementApiServiceImpl(
...
+      _ <- validateResultState(newState)
+      _ <- authorizationManagementService.updateStateOnClients(
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: after sleeping on it, should this be performed only if the new state is `ACTIVE` or `SUSPENDED`?
If the new agreement is not in a "operational" state, clients should keep the current configuration (for example if the resulting state is `PENDING`).
Does this make sense?

Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: Hmm, indeed it does, that's probably was was there in the first place

Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: Updated

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 6]
```diff
@@ -178,24 +181,50 @@ final case class AgreementApiServiceImpl(
...
+            descriptorId = newDescriptorId,
+            producerId = agreement.producerId,
+            consumerId = agreement.consumerId,
+            verifiedAttributes = agreement.verifiedAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            certifiedAttributes = agreement.certifiedAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            declaredAttributes = agreement.declaredAttributes.map(a => AgreementManagement.AttributeSeed(a.id)),
+            consumerNotes = agreement.consumerNotes
+          )
+        )
+      } yield agreement
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: The `consumerDocuments` should be copied too.
You can use the function `createAndCopyDocumentsForClonedAgreement` (see the comment below).

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 7]
```diff
@@ -222,7 +251,10 @@ final case class AgreementApiServiceImpl(
       eService       <- catalogManagementService.getEServiceById(agreement.eserviceId)
       _              <- verifyCloningConflictingAgreements(requesterOrgId, agreement.eserviceId)
       consumer       <- tenantManagementService.getTenant(requesterOrgId)
-      _              <- validateCertifiedAttributes(eService, consumer).whenA(eService.producerId != consumer.id)
+      descriptor     <- eService.descriptors
+        .find(_.id == agreement.descriptorId)
+        .toFuture(DescriptorNotFound(eService.id, agreement.descriptorId))
+      _              <- validateCertifiedAttributes(descriptor, consumer).whenA(eService.producerId != consumer.id)
       newAgreement   <- agreementManagementService.createAgreement(agreement.toSeed)
       documents      <- createAndCopyDocumentsForClonedAgreement(agreement, newAgreement)
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala:260]: [here](https://github.com/pagopa/interop-be-agreement-process/pull/233/files#r1239525784)

File: project/Versions.scala (hunk 1 above)
Comment (beetlecrunch) [reply project/Versions.scala]: I am not alone 😄

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala (hunk 6 above)
Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: good catch

Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: I would rather prefer to make it "more" atomic, and modify the AgreementSeed in both management and process, wdyt?

Comment (beetlecrunch) [reply src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: The new field will be passed nonempty only in this specific case.

**Review by beetlecrunch — APPROVED** [review]

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala [hunk 8]
```diff
@@ -62,6 +62,7 @@ object ResponseHandlers extends AkkaResponses {
     result match {
       case Success(s)                               => success(s)
       case Failure(ex: AgreementNotFound)           => notFound(ex, logMessage)
+      case Failure(ex: DescriptorNotFound)          => badRequest(ex, logMessage)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala]: IMO this should return 500

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala [hunk 9]
```diff
@@ -73,6 +74,7 @@ object ResponseHandlers extends AkkaResponses {
...
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala]: same

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala [hunk 10]
```diff
@@ -121,6 +123,7 @@ object ResponseHandlers extends AkkaResponses {
       case Failure(ex: AgreementNotFound)           => notFound(ex, logMessage)
       case Failure(ex: OperationNotAllowed)         => forbidden(ex, logMessage)
       case Failure(ex: AgreementNotInExpectedState) => badRequest(ex, logMessage)
+      case Failure(ex: DescriptorNotFound)          => badRequest(ex, logMessage)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/ResponseHandlers.scala]: same

File: src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala [hunk 11]
```diff
@@ -625,21 +630,28 @@ final case class AgreementApiServiceImpl(
...
         if (updated.state == AgreementManagement.AgreementState.ACTIVE && agreements.isEmpty)
           createContractAndSendMail(updated, updateSeed)
         else Future.unit
     } yield updated
+  }
 
+  def activeOrSuspended(newState: AgreementManagement.AgreementState): Boolean = newState match {
+    case AgreementManagement.AgreementState.ACTIVE | AgreementManagement.AgreementState.SUSPENDED => true
+    case _                                                                                        => false
   }
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/agreementprocess/api/impl/AgreementApiServiceImpl.scala]: just minor consideration
```suggestion
  def activeOrSuspended(newState: AgreementManagement.AgreementState): Boolean =
    newState == AgreementManagement.AgreementState.ACTIVE || newState == AgreementManagement.AgreementState.SUSPENDED 
```

**Review by galales — APPROVED** [review]
LGTM, I would just let return 500 in the commented cases

**Pull request merged**